### ISSUE TITLE
I-JEPAの性能検証実験

### ISSUE THREAD

**Issue by Geson-anko** [MLShukai/ami-vconf24#94]
# 実験内容

## 概要

I-JEPAの動作検証実験を行う。ハイパラのチューニング等を行い、観測画像が適切にエンコードできることを確かめる。
実験はVRChatのJapan Streetをランダム行動で動かした観測の記録を**24時間**分用いて行う。
ハイパーパラメータのチューニングは主に 学習率に対して行うことになるだろう。

## 方針

* 事前に記録した観測動画を用いて I-JEPAを学習させる。
* 損失が低下することを確認するほか、Context EncoderやTarget Encoderの出力の標準偏差がある程度保たれることを確認する。
* Context Encoder, Target Encoderどちらも再構成画像を作り、観測情報を十分に保持できているか目で見て確かめる。
* 本家I-JEPAのPaperで報告されたもの（再構成画像や損失の値）と比較し、検証する。
  **TODO: どのように比較する？**

### 手順
実験用計算機を用いて事前に収集した動画で学習する手順を示す。

1. 計算機にsshでログインする。ユーザ名は`gop-geson`である。参考: https://drive.google.com/drive/folders/1CFD0AFiHxyucMcszh8LjMFikcx6t83rL?usp=share_link
2. 実験環境に入る（立ち上げる） 
    1. コマンドを実行`cd /mnt/Data/ami-vconf24`  
    2. `make docker-attach`  
    NOTE: 失敗した場合は、`make docker-run-with-data DATA_DIR=/mnt/Archive`を実行してから 再度実行。
3. 実験ブランチに切り替える。  
    1. `git checkout experiment/#94/observation-encoder-i-jepa`  
    2. `git pull`  
4. `tmux`コマンドを実行。  
    端末が途切れても実行され続けるようにするため。
5. 実験を実行する。
    ```bash
    python scripts/launch.py experiment=bool_mask_i_jepa_with_videos max_uptime=86400 # 24 hours
    ```
    `logs/bool_mask_i_jepa_with_videos/runs`に出力される。

#### 実験オプション

コマンドラインオプションを追加することでハイパーパラメータ等を変更できる。

* 時間スケール `time_scale=<float>`
  AMIの実行速度を変化させられる。（計算機の性能依存)  
  10fpsで通常は動作するが、Largeサイズの単一実行なら`1.4`, 2並列で実行する際は `0.7`程度の値がちょうど良かった。
* 最大稼働時間 `max_uptime=<float>`
  最大実行時間。この時間に到達すると終了する。`time_scale`の値でスケールされる。
* 複数実行 multirun `hydra.mode=MULTIRUN option=value1,value2`
  複数のハイパーパラメータを同時実行できる。  
  `hydra.launcher.n_jobs=<unsigned int>`をさらに追加することで、並列実行する数を指定できる。
  **NOTE**: multirunの時は `logs/<task_name>/multiruns`にログが出力される。
  **NOTE**: 複数並列で動作させると実行時間がそれに比例して長くなるので`time_scale`の設定に注意
* その他のコンフィグパラメータの上書き
    全てのハイパーパラメータは、configの階層構造 (`configs`ディレクトリ構造)に従ってコマンドラインから上書きできる。  
    詳しくは次の記事を参考にしてほしい。 https://zenn.dev/gesonanko/articles/417d43669cf2af  
    
    学習率を変える例
    ```py
    python ... trainers.i_jepa.partial_optimizer.lr=<value>
    ```
    
    **tabキーを押すことで補完する機能があるのでぜひ試して欲しい**

### 実験結果の確認方法

Tensorboardを起動して確認する。
```bash
tensorboard --bind_all --logdir ./logs/bool_mask_i_jepa_with_videos
```

Webブラウザからアクセスする。 `<計算機のip>:6006`からアクセスすることができる。

起動時のHyper parameterは `HPARAMS`タブから確認することができるが、見にくい。 ログディレクトリの `main.log`ファイルにconfigファイル全体が出力されているのでそれを見にいく方が見やすい。

### モデル

モデルサイズ Large, Medium, Smallの内どれかを選択する。(2024/08/30 デフォルト Large)
起動時の引数に `models=bool_mask_i_jepa_<size>` を指定する。(`<size>`はlarge, medium, smallのどれかに置き換えてください)

###  Trainer

`configs/trainers/i_jepa/bool_mask.yaml`の内容。

* `mask_scale`
  マスク1枚あたりのサイズの範囲。 `[min, max]`という形式で指定する。デフォルトでは `[0.1, 0.25]`なっている。(1/10 ~ 1/4 程度を隠す)
  ただ、実際には `n_masks = 4` 枚サンプルされ、それを合わせたもので全体のマスクが形成される。そのため、マスクされない領域の期待値は $0.463 = (1 - \frac{0.1 + 0.25}{2})^4$  となっている。  
  本家I-JEPAはマスクされない領域サイズの期待値は $0.656$と計算されるたため、デフォルト値は少々大きすぎる可能性が高い。
  本系に合わせると、 マスクサイズは `[0.025, 0.125]`, マスクされない領域の期待値が $0.656 = (1 - \frac{0.025 + 0.125}{2})^4$ とすると良いかしれない。

* `min_keep`
  マスクしない最小領域パッチ数。 デフォルトで`10`とされている。
  I-JEPAの本家では パッチ数が $196 = 14 \times 14$であった。今回の実装ではパッチ数は $144 = 12 \times 12$であるため、それと同じ割にするならば `6`とすることが望ましいだろう。

* 学習率 `lr`
  デフォルトでは `0.0001`となっている。学習が安定しないようであれば下げることが望ましい。
  P-AMI<Q>のlifetimeは **1日**としているため、その期間学習が安定して行われる値にチューニングする。
  変化させる際は $[10^{-3}, 10^{-5}]$の範囲から指数関数的にサンプルすると良い (Ex. `1e-3, 5e-4, 2e-4, 1e-4, 8e-5, 5e-5, 2e-5, 1e-5`)

* `target_encoder_update_moving_average`
  Context EncoderからTarget Encoderへ指数移動平均を用いて重みをコピーしていく際の1更新あたりの割合。デフォルト値は 本家I-JEPAの初期値に合わせ`0.996` (0.4%更新される)としている。小さな値にすると (0.9など)学習が崩壊する。大きな値にするとConext Encoderのターゲットデータが安定して生成されるようになるため、Context Encoderの学習が安定すると予想される。  
  本家 I-JEPAでは 0.996から1.0へ学習が終了するまで線形に増加させていた。P-AMI<Q>において学習終了はないためそれは行わない。  

### 必ず行いたいこと

* マスクサイズを  `[0.025, 0.125]`にした場合の変化 
    `trainers.i_jepa.partial_dataloader.collate_fn.mask_scale=[0.025,0.125]`
* モデルサイズをMedium, Smallにした場合の変化  
    `models=bool_mask_i_jepa_large,bool_mask_i_jepa_medium,bool_mask_i_jepa_small`
* 最適な学習率の検索 (lr=0.001, 0.0001, 0.00001は少なくとも調べよう)
    `trainers.i_jepa.partial_optimizer.lr=1e-3,1e-4,1e-5`
* 最適な`target_encoder_update_moving_average`
    `trainers.i_jepa. target_encoder_update_moving_average=0.96,0.996,0.9996,0.99996`

## 評価項目

本家I-JEPAの再構成画像と比較して、主観的に綺麗かどうか。
また、再構成誤差が低下し続けるか。ValidationデータとTrainingデータとの誤差の乖離が少ないほど良い。

## 参考

実験設計について:  https://docs.google.com/document/d/1VeUSs2rR0cL4gncGO38X_FR9WWaaAN_NxsOcjCqFz-Q/edit?usp=sharing

Hydraについて: https://zenn.dev/gesonanko/articles/417d43669cf2af

**Comment by zassou65535** [comment]
学習率`lr`だけど、I-JEPAの論文の巻末にあるA. Implementation DetailsのOptimizationに学習中の`lr`の細かい変化について記述されているからそこも参考にしてみるといいかもしれない

**Comment by Geson-anko** [comment]
@zassou65535 
> Optimization. We use AdamW [51] to optimize the context-encoder and predictor weights. Our default batch-size is 2048, and the learning rate is linearly increased from 10−4 to 10−3 during the first 15 epochs of pretraining, and decayed to 10−6 following a cosine schedule thereafter. Following [4, 18], the weight-decay is linearly increased from 0.04 to 0.4 throughout pretraining. The target-encoder weights are identical to the context-encoder weights at initialization, and updated via an exponential moving average thereafter [4, 18, 23, 35, 37, 61]. We use a momentum value of 0.996, and linearly increase this value to 1.0 throughout pretraining, following [4, 18].

最初Warmupが行われているみたいだね。もしイマイチ性能が上がらない見たいなら初期の学習率スケジューリングを考えてもいいかもね。
